redis/src/cluster_client.rs

```diff
@@ -1,164 +1,186 @@
 use crate::cluster::ClusterConnection;
+use crate::connection::{ConnectionAddr, ConnectionInfo, IntoConnectionInfo};
+use crate::types::{ErrorKind, RedisError, RedisResult};
 
-use super::{
-    ConnectionAddr, ConnectionInfo, ErrorKind, IntoConnectionInfo, RedisError, RedisResult,
-};
+/// Redis cluster specific parameters.
+#[derive(Default, Clone)]
+pub(crate) struct ClusterParams {
+    pub(crate) password: Option<String>,
+    pub(crate) username: Option<String>,
+    pub(crate) read_from_replicas: bool,
+}
 
 /// Used to configure and build a [ClusterClient](ClusterClient).
 pub struct ClusterClientBuilder {
     initial_nodes: RedisResult<Vec<ConnectionInfo>>,
-    read_from_replicas: bool,
-    username: Option<String>,
-    password: Option<String>,
+    cluster_params: ClusterParams,
 }
 
 impl ClusterClientBuilder {
-    /// Generate the base configuration for new Client.
-    pub fn new<T: IntoConnectionInfo>(initial_nodes: Vec<T>) -> ClusterClientBuilder {
+    /// Creates a new `ClusterClientBuilder` with the the provided initial_nodes.
+    ///
+    /// This is the same as `ClusterClient::builder(initial_nodes)`.
+    pub fn new<T>(initial_nodes: Vec<T>) -> ClusterClientBuilder
+    where
+        T: IntoConnectionInfo,
+    {
         ClusterClientBuilder {
             initial_nodes: initial_nodes
                 .into_iter()
                 .map(|x| x.into_connection_info())
                 .collect(),
-            read_from_replicas: false,
-            username: None,
-            password: None,
+            cluster_params: ClusterParams::default(),
         }
     }
 
-    /// Builds a [ClusterClient](ClusterClient). Despite the name, this does not actually open
-    /// a connection to Redis Cluster, but will perform some basic checks of the initial
-    /// nodes' URLs and passwords.
+    /// Creates a new `ClusterClient` with the parameters.
+    ///
+    /// This does not create connections to the Redis Cluster, but only performs some basic checks
+    /// on the initial nodes' URLs and passwords/usernames.
     ///
     /// # Errors
     ///
-    /// Upon failure to parse initial nodes or if the initial nodes have different passwords,
-    /// an error is returned.
-    pub fn open(self) -> RedisResult<ClusterClient> {
-        ClusterClient::build(self)
+    /// Upon failure to parse initial nodes or if the initial nodes have different passwords or
+    /// usernames, an error is returned.
+    pub fn build(self) -> RedisResult<ClusterClient> {
+        let mut initial_nodes = self.initial_nodes?;
+
+        let mut cluster_params = self.cluster_params;
+        let password = if cluster_params.password.is_none() {
+            cluster_params.password = initial_nodes[0].redis.password.clone();
```

Comment:
This will cause a panic if `initial_nodes` is empty for some reason.

Reply:
So this invariant is not checked here. Is it checked somewhere else? If yes, then we should add a comment pointing that out. If not, should check it here.

Reply:
Ok.